### PR TITLE
[SPARK-31076][SQL][FOLLOWUP] Incapsulate date rebasing to `DaysWritable`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Move the code related to days rebasing from/to Julian calendar from `HiveInspectors` to new class `DaysWritable`.

### Why are the changes needed?
To improve maintainability of the `HiveInspectors` trait which is already pretty complex.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
By `HiveOrcHadoopFsRelationSuite`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon Please, take a look at this.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala [hunk 1]
```diff
@@ -646,14 +617,7 @@ private[hive] trait HiveInspectors {
...
             if (data != null) {
-              // Rebasing written days via conversion to local dates.
-              // See the comment for `getDateWritable()`.
-              val daysSinceEpoch = x.getPrimitiveWritableObject(data).getDays
-              if (daysSinceEpoch < JULIAN_CUTOVER_DAY) {
-                rebaseJulianToGregorianDays(daysSinceEpoch)
-              } else {
-                daysSinceEpoch
-              }
+              x.getPrimitiveWritableObject(data).getDays
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: What if the returned `DateWritable` here is not our custom version?

Comment (MaxGekk) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: You are right, `DateWritable` can come here. I have fixed that by adding new constructor to `DaysWritable` to construct its instances from `DateWritable`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/DaysWritable.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  def this(dateWritable: DateWritable) = {
+    this(dateWritable match {
+      case daysWritable: DaysWritable => daysWritable.gregorianDays
+      case dateWritable: DateWritable =>
+        DaysWritable.rebaseJulianToGregorianDays(dateWritable.getDays)
+    })
+  }
+
+  override def getDays: Int = gregorianDays
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/DaysWritable.scala]: should we return `julianDays` here to conform to the hive API?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/DaysWritable.scala [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    // as by Spark 2.4 and earlier versions. Spark 3.0 switched to
+    // Proleptic Gregorian calendar (see SPARK-26651), and as a consequence of that,
+    // this affects dates before 1582-10-15. Spark 2.4 and earlier versions use
+    // Julian calendar for dates before 1582-10-15. So, the same local date may
+    // be mapped to different number of days since the epoch in different calendars.
+    // For example:
+    // Proleptic Gregorian calendar: 1582-01-01 -> -141714
+    // Julian calendar: 1582-01-01 -> -141704
+    // The code below converts -141714 to -141704.
+    val writeDays = DaysWritable.rebaseGregorianToJulianDays(gregorianDays)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/DaysWritable.scala]: shall we cache the `julianDays`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/DaysWritable.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/DaysWritable.scala]: done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/DaysWritable.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/DaysWritable.scala]: done

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by HyukjinKwon** [comment]
+1 Looks good to me too

**Pull request closed without merging**